File: go/analytics.go
package youtube

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

// DefaultAnalyticsBaseURL is the production YouTube Analytics API v2 root.
const DefaultAnalyticsBaseURL = "https://youtubeanalytics.googleapis.com/v2"

// AnalyticsColumnHeader describes one column in an analytics report.
type AnalyticsColumnHeader struct {
	Name       string `json:"name"`
	ColumnType string `json:"columnType"`
	DataType   string `json:"dataType"`
}

// AnalyticsReport is the response from an analytics query. Rows is a matrix of
// untyped JSON values whose order matches ColumnHeaders; callers interpret each
// cell per its column's DataType.
type AnalyticsReport struct {
	Kind          string                  `json:"kind"`
	ColumnHeaders []AnalyticsColumnHeader `json:"columnHeaders"`
	Rows          [][]json.RawMessage     `json:"rows"`
}

// AnalyticsQueryParams configures AnalyticsQuery.
type AnalyticsQueryParams struct {
	// IDs identifies the channel or content owner. Defaults to
	// "channel==MINE".
	IDs string
	// StartDate / EndDate are inclusive YYYY-MM-DD bounds. Required.
	StartDate string
	EndDate   string
	// Metrics is the list of metrics to report (e.g. ["views","likes"]).
	// Required.
	Metrics []string
	// Dimensions is the optional list of dimensions (e.g. ["day"]).
	Dimensions []string
	// Filters is an optional semicolon-delimited filter expression.
	Filters string
	// Sort is the optional list of sort fields ("-" prefix for descending).
	Sort []string
	// MaxResults limits the number of rows (0 = API default).
	MaxResults int
	// StartIndex is the 1-based index of the first row (0 = API default).
	StartIndex int
	// Currency is the ISO 4217 code for revenue metrics.
	Currency string
}

// AnalyticsQuery runs a YouTube Analytics report query. Requires an OAuth
// access token (the Analytics API does not accept API keys).
func (c *Client) AnalyticsQuery(ctx context.Context, p AnalyticsQueryParams) (*AnalyticsReport, error) {
	if err := c.requireOAuth(); err != nil {
		return nil, err
	}
	if p.StartDate == "" || p.EndDate == "" {
		return nil, errors.New("youtube: AnalyticsQuery requires StartDate and EndDate")
	}
	if len(p.Metrics) == 0 {
		return nil, errors.New("youtube: AnalyticsQuery requires at least one metric")
	}

	ids := p.IDs
	if ids == "" {
		ids = "channel==MINE"
	}

	q := url.Values{}
	q.Set("ids", ids)
	q.Set("startDate", p.StartDate)
	q.Set("endDate", p.EndDate)
	q.Set("metrics", strings.Join(p.Metrics, ","))
	if len(p.Dimensions) > 0 {
		q.Set("dimensions", strings.Join(p.Dimensions, ","))
	}
	if p.Filters != "" {
		q.Set("filters", p.Filters)
	}
	if len(p.Sort) > 0 {
		q.Set("sort", strings.Join(p.Sort, ","))
	}
	if p.MaxResults > 0 {
		q.Set("maxResults", strconv.Itoa(p.MaxResults))
	}
	if p.StartIndex > 0 {
		q.Set("startIndex", strconv.Itoa(p.StartIndex))
	}
	if p.Currency != "" {
		q.Set("currency", p.Currency)
	}

	out := &AnalyticsReport{}
	if err := c.getAnalytics(ctx, "reports", q, out); err != nil {
		return nil, err
	}
	return out, nil
}

// getAnalytics issues a GET against the Analytics API host. The Analytics base
// URL is fixed (it is not overridable via ClientOptions today); tests exercise
// AnalyticsQuery via the request-building path with a stubbed transport.
func (c *Client) getAnalytics(ctx context.Context, path string, query url.Values, out any) error {
	base := c.analyticsBaseURL()
	fullURL := base + "/" + strings.TrimLeft(path, "/") + "?" + query.Encode()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fullURL, nil)
	if err != nil {
		return fmt.Errorf("youtube: build analytics request: %w", err)
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Authorization", "Bearer "+c.accessToken)
	req.Header.Set("User-Agent", c.userAgent)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("youtube: analytics request: %w", err)
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("youtube: read analytics response: %w", err)
	}
	if resp.StatusCode >= 400 {
		return parseErrorBody(resp.StatusCode, raw)
	}
	if out == nil || len(raw) == 0 {
		return nil
	}
	if err := json.Unmarshal(raw, out); err != nil {
		return &Error{
			StatusCode: resp.StatusCode,
			Status:     "INVALID_RESPONSE",
			Message:    fmt.Sprintf("analytics returned non-JSON: %v", err),
			Body:       raw,
		}
	}
	return nil
}

// analyticsBaseURL returns the Analytics host, honoring an override stored on
// the client for tests.
func (c *Client) analyticsBaseURL() string {
	if c.analyticsBase != "" {
		return c.analyticsBase
	}
	return DefaultAnalyticsBaseURL
}

File: youtube/apis/analytics.py
"""YouTube Analytics API namespace.

Wraps the following endpoint:

* GET /reports — query analytics reports

Reference: https://developers.google.com/youtube/analytics/reference/reports/query
"""

from __future__ import annotations

from typing import Any

from youtube.apis.base import BaseAPI
from youtube.config import YOUTUBE_ANALYTICS_BASE_URL
from youtube.models.analytics import AnalyticsResponse

_BASE = YOUTUBE_ANALYTICS_BASE_URL


class AnalyticsAPI(BaseAPI):
    """Methods for querying YouTube Analytics reports."""

    async def query(
        self,
        *,
        ids: str = "channel==MINE",
        start_date: str,
        end_date: str,
        metrics: str | list[str],
        dimensions: str | list[str] | None = None,
        filters: str | None = None,
        sort: str | list[str] | None = None,
        max_results: int | None = None,
        start_index: int | None = None,
        currency: str | None = None,
        include_historical_channel_data: bool | None = None,
    ) -> AnalyticsResponse:
        """Query a YouTube Analytics report.

        Parameters
        ----------
        ids:
            Channel or content owner to report on.
            Default ``"channel==MINE"`` uses the authenticated user's channel.
        start_date:
            Start date in ``YYYY-MM-DD`` format.
        end_date:
            End date in ``YYYY-MM-DD`` format.
        metrics:
            Comma-separated metrics or a list, e.g.
            ``["views", "likes", "estimatedMinutesWatched"]``.
        dimensions:
            Comma-separated dimensions or a list, e.g. ``["day"]``.
        filters:
            Semicolon-separated filters, e.g.
            ``"video==VIDEO_ID;country==US"``.
        sort:
            Comma-separated sort fields or a list.  Prefix ``-`` for
            descending.
        max_results:
            Maximum number of rows to return.
        start_index:
            1-based index of the first row to return.
        currency:
            ISO 4217 currency code for revenue metrics (default ``USD``).
        include_historical_channel_data:
            Include data from before the channel was linked.
        """
        params: dict[str, Any] = {
            "ids": ids,
            "startDate": start_date,
            "endDate": end_date,
            "metrics": ",".join(metrics) if isinstance(metrics, list) else metrics,
        }
        if dimensions is not None:
            params["dimensions"] = (
                ",".join(dimensions) if isinstance(dimensions, list) else dimensions
            )
        if filters is not None:
            params["filters"] = filters
        if sort is not None:
            params["sort"] = ",".join(sort) if isinstance(sort, list) else sort
        if max_results is not None:
            params["maxResults"] = max_results
        if start_index is not None:
            params["startIndex"] = start_index
        if currency is not None:
            params["currency"] = currency
        if include_historical_channel_data is not None:
            params["includeHistoricalChannelData"] = str(include_historical_channel_data).lower()

        payload = await self._session.get(f"{_BASE}/reports", params=params)
        return AnalyticsResponse.model_validate(payload)
